Issue610 cdl additional blocks

milicag opened the pull request:
The SetPoints package now has the following blocks:
- CalenderTime
- HotWaterTemperatureReset
- ModelTime
- CombiTimeTable

These should, for the time being, cover the input requirements. I've replaced ```Table.mo``` with one listed above, due to better functionality.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/SetPoints/CalendarTime.mo:
```diff
@@ -0,0 +1,360 @@
+    "Enumeration for choosing how reference time (time = 0) should be defined";
+  parameter Integer yearRef(min=firstYear, max=lastYear) = 2016
+    "Year when time = 0, used if zerTim=Custom"
+    annotation(Dialog(enable=zerTim==Buildings.Utilities.Time.Types.ZeroTime.Custom));
```
`Buildings.Utilities.Time.Types.ZeroTime` also needs to be copied to `OpenBuildingControl.CDL` so that the code is self-contained.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/SetPoints/CalendarTime.mo:
```diff
@@ -0,0 +1,360 @@
+    defaultComponentName="calTim",
+  Documentation(revisions="<html>
+<ul>
+<li>
```
Please update release notes

milicag replied:
I've updated them, but the pull request is not displaying at the moment.

mwetter replied:
@milicag If the code does not show in the pull request, you probably either did not push it, or committed to another branch?

milicag replied:
Updated

mwetter replied:
I usually close Dymola if I need to make sure that all files are saved. Dymola also has a function to save all files, but this, at least in the past, lead to small formatting changes in many files which made version control difficult.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/SetPoints/CombiTimeTable.mo:
```diff
@@ -0,0 +1,524 @@
+  parameter Integer columns[:]=2:size(table, 2)
+    "Columns of table to be interpolated"
+    annotation (Dialog(group="Table data interpretation"));
+  parameter Modelica.Blocks.Types.Smoothness smoothness=Modelica.Blocks.Types.Smoothness.LinearSegments
```
Types need to be moved to `CDL`

milicag replied:
At some point we've had a discussion with a conclusion to retain any types from Modelica library for now. There are a bunch of such types at other places, e.g. ```Modelica.SIunits.Time```, several others in Psychrometrics. Let me know if you'd like to reconsider that.

mwetter replied:
@milicag : enumerations such as `Modelica.Blocks.Types.Smoothness` should be moved to `CDL`, but types used for units we can leave (for now) in `Modelica.SIunits`. We may move them later if needed.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/SetPoints/ModelTime.mo:
```diff
@@ -0,0 +1,51 @@
+within Buildings.Experimental.OpenBuildingControl.CDL.SetPoints;
+block ModelTime "Model time"
```
This should rather be called 'StandardTime' as I don't think it is meaningful to talk about 'ModelTime' in a building automation system.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/SetPoints/ModelTime.mo:
```diff
@@ -0,0 +1,51 @@
+          points={{0,0},{40,0}},
+          thickness=0.5)}),
+    Documentation(info="<html>
+<p>This component outputs the model time, which starts at the value at which the simulation starts. For example, if a simulation starts at <i>t=-1</i>, then this block outputs first <i>t=-1</i>, and its output is advanced at the same rate as the simulation time. </p>
```
use shorter lines.

mwetter commented on Buildings/Controls/SetPoints/Table.mo:
```diff
@@ -27,7 +27,7 @@ protected
 equation
   connect(u, tab.u[1]) annotation (Line(
-      points={{-120,0},{-70,0},{-70,0},{-22,0}},
+      points={{-120,0},{-70,0},{-70,0},{-22.2,0}},
```
Please revert this change.

milicag replied:
I got rid of the ```Table.mo``` and added ```CombiTimeTable.mo``` for the sake of it's better features.

milicag replied:
I'm having issues with dymola not updating the file if edited and saved in the dymola editor. Here is a screen shot showing the issue (see line 30 both in dymola and in the text editor). After I introduced the change in the text editor, the changes were saved and recorded by git.
![screenshot from 2017-02-27 13-29-51](https://cloud.githubusercontent.com/assets/5322943/23380891/5f11186e-fcf1-11e6-9d07-e454efb0bbf9.png)

milicag commented:
I updated based on your comments, so it should be ready for review. Apart from the easy details, it's worth noting that I've replaced ```Table.mo``` with the ```CombiTimeTable.mo```. Please also see the note on retaining any references to modelica blocks that are not extend statements, as discussed at some point.

mwetter commented on Buildings/Experimental/OpenBuildingControl/CDL/Types/ZeroTime.mo:
```diff
@@ -0,0 +1,45 @@
+</html>", revisions="<html>
+<ul>
+<li>
+September 10, 2016, by Michael Wetter:<br/>
```
Revision notes need to be updated

mwetter commented:
@milicag : Also, the html validation fails, please see the above failed continuous integration test.

milicag commented:
I'm waiting for Travis to run the tests. Travis runs tests automatically for each commit?

milicag commented:
I'm still getting the following html error:
```
The following malformed html syntax has been found:

[-- ./Experimental/OpenBuildingControl/CDL/SetPoints/CombiTimeTable.mo ]

line 29 column 1 - Warning: <img> element not empty or not closed

line 173 column 88 - Warning: <br> element not empty or not closed

line 325 column 76 - Warning: <br> element not empty or not closed

line 449 column 27 - Warning: <br> element not empty or not closed

line 455 column 78 - Warning: <br> element not empty or not closed
```

I was not able to find any obvious differences compared to the original file (```Modelica.Blocks.Sources.CombiTimeTable```), but I've made some edits and I'm waiting for the tests to complete again.

mwetter commented:
@milicag 
Regarding the html errors, 

* the `<img>` tag needs to be closed with `<img .... />`, see for example https://github.com/lbl-srg/modelica-buildings/blob/a820d18ca54e060295ea949ab3269d36b0bbe031/Buildings/package.mo#L6442
* the other problems are because line breaks are `<br/>`, not `<br>`.

You can also run the tests locally with
```
cd Buildings
../bin/runUnitTests.py --validate-html-only
```

milicag commented:
Thanks for the shortcut to run the html tests. I'll need to install tidylib, which prompted me to look for ```requirements.txt```. Do you guys have it in the repo?

milicag commented:
This should be a functional version. I'm a bit concerned about the ```Modelica.Blocks.Types.ExternalCombiTimeTable``` type in the ```CombiTimeTable```. I tried to duplicate it in CDL, but I'm running into this error:

```
/opt/dymola-2017-x86_64/source/matrixop.c: In function ‘Integer2String’:
/opt/dymola-2017-x86_64/source/matrixop.c:3579:14: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 5 has type ‘int’ [-Wformat=]
  sprintf(buf,"%*.*ld",(int)minwidth,(int)precision,(int)x);
              ^
/opt/dymola-2017-x86_64/source/matrixop.c: In function ‘Integer2String2’:
/opt/dymola-2017-x86_64/source/matrixop.c:3600:28: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 4 has type ‘int’ [-Wformat=]
  sprintf(buf,leftJustified?"%-*ld":"%*ld",(int)minwidth,(int)x);
                            ^
/opt/dymola-2017-x86_64/source/matrixop.c:3600:36: warning: format ‘%ld’ expects argument of type ‘long int’, but argument 4 has type ‘int’ [-Wformat=]
  sprintf(buf,leftJustified?"%-*ld":"%*ld",(int)minwidth,(int)x);
                                    ^
In file included from /opt/dymola-2017-x86_64/source/matrixop.c:3746:0:
/opt/dymola-2017-x86_64/source/localeless.cpp: In function ‘sscanfCx2_underscore’:
/opt/dymola-2017-x86_64/source/localeless.cpp:308:22: warning: format ‘%lx’ expects argument of type ‘long unsigned int *’, but argument 3 has type ‘unsigned int *’ [-Wformat=]
  return sscanf(line, "%lx_%lx", a,b);
                      ^
/opt/dymola-2017-x86_64/source/localeless.cpp:308:22: warning: format ‘%lx’ expects argument of type ‘long unsigned int *’, but argument 4 has type ‘unsigned int *’ [-Wformat=]
/opt/dymola-2017-x86_64/source/localeless.cpp: In function ‘sscanfClu’:
/opt/dymola-2017-x86_64/source/localeless.cpp:319:22: warning: format ‘%lu’ expects argument of type ‘long unsigned int *’, but argument 3 has type ‘unsigned int *’ [-Wformat=]
  return sscanf(line, "%lu", a);
                      ^
/usr/bin/ld: cannot find -lModelicaMatIO
/usr/bin/ld: cannot find -lModelicaStandardTables
/usr/bin/ld: cannot find -lzlib
/usr/bin/ld: cannot find -lModelicaMatIO
/usr/bin/ld: cannot find -lModelicaStandardTables
/usr/bin/ld: cannot find -lzlib
collect2: error: ld returned 1 exit status
```

milicag commented:
This should be ready for review, if the Travis tests pass. As discussed, I've left the modelica library type ```Modelica.Blocks.Types.ExternalCombiTimeTable``` in as a protected parameter.
